### ISSUE TITLE
Duplicate symbols when building GenerateFingerPrintTables

### ISSUE THREAD

**Issue by JoeLoser** [facebook/folly#748]
With Clang 5.0.1, CMake 3.9.6, and Mac OS version 10.13.12, I am seeing duplicate symbols resulting in linker errors when building the target `GenerateFingerPrintTables`. This is when using an out-of-tree build (e.g. running `CMake` from a build directory).

```
duplicate symbol __ZN5folly6detail12formatBinaryE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail14formatHexLowerE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail14formatHexUpperE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail11formatOctalE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail16formatAlignTableE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail15formatSignTableE in:
    CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o
duplicate symbol __ZN5folly6detail8hexTableE in:
    CMakeFiles/folly_base.dir/folly/EscapeTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/EscapeTables.cpp.o
duplicate symbol __ZN5folly6detail14cUnescapeTableE in:
    CMakeFiles/folly_base.dir/folly/EscapeTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/EscapeTables.cpp.o
duplicate symbol __ZN5folly6detail14uriEscapeTableE in:
    CMakeFiles/folly_base.dir/folly/EscapeTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/EscapeTables.cpp.o
duplicate symbol __ZN5folly6detail12cEscapeTableE in:
    CMakeFiles/folly_base.dir/folly/EscapeTables.cpp.o
    CMakeFiles/folly_base.dir/folly/build/EscapeTables.cpp.o
ld: 10 duplicate symbols for architecture x86_64
```

@simpkins I think it might be related to your change in revision 3d9eb7ffc32c2f5f878bb4aee61d30cb8c62a146? Do you have any insights into this? Thanks!

**Comment by simpkins** [comment]
I'm surprised you have files located at CMakeFiles/folly_base.dir/folly/FormatTables.cpp.o
I would only expect the ones under folly/build (e.g., CMakeFiles/folly_base.dir/folly/build/FormatTables.cpp.o); I'm not sure how or why the build would also end up generating these files one directory above too.

The CMake build hasn't really been tested on Mac much yet, so it wouldn't necessarily surprise me if there are some issues that still need to be ironed out.  (The CMake build used to only work on Windows; It wasn't until the last couple weeks that we started making it work on other platforms, but I've mainly only been testing on Linux distributions.)  If you're able to track down the issues and submit a pull request we'll likely accept it.

That said, @yfeldblum also has some commits that will land shortly to remove the separate build steps for EscapeTables, FormatTables, and GroupVarintTables.  I believe FingerprintTables will be the only file left that will still be generated by a separate build rule.

**Comment by simpkins** [comment]
Ah, that makes sense.

In the long run I would ideally like to change the CMake build so that it uses an explicit file list instead of its current glob-based scheme to find all the files and then exclude specific ones.  That would help avoid confusing situations like this.  However, I don't really plan to tackle this myself anytime soon.  (We would likely want to build some mechanism to keep the file lists in sync between our internal build files and the CMake build files.)

**Comment by JoeLoser** [comment]
Yeah, explicit file listing for targets is really the golden solution I've found from my experience. I've seen both worlds on mid-sized and large projects internally at my work. One thing that I find a bit unintuitive/not ideal is remembering to reload CMake when you add/delete a file. Using a globbing approach, that filename would exist in `CMakeCache` but not in your tree (once you've deleted it for example) -- forcing the reload/regeneration.

CMake's philosophy is that you should be able to *generate* the `CMakeCache` and build directory once and do *many* builds for it while relying on your build tool `make `, `ninja`, etc to determine whether a rebuild is necessary. As such, I've found explicit file listings to have fewer surprises associated with them. 

Not sure I follow the last part about keeping the file lists in sync between internal and open-source world-- the header files and cpp files or the CMake files?
